lib/gruf/server.rb

```diff
@@ -64,44 +64,54 @@ def start!
     end
     # :nocov:
 
-    private
-
     ##
-    # Return all loaded gRPC services
+    # @param [Class] klass
     #
-    # @param [Array<Class>] svcs An array of service classes that will be handled by this server
-    # @return [Array<Class>] The given services that were added
-    #
-    def load_services(svcs)
-      unless @services
-        @services = Gruf.services.concat(svcs)
-        @services.uniq!
-      end
-      @services
+    def add_service(klass)
+      @services << klass unless @services.include?(klass)
     end
 
     ##
+    # @param
+    def add_interceptor(klass, opts = {})
+      @interceptors[klass.name.to_s] = opts
+    end
+
+    private
+
+    ##
     # Auto-load all gRPC handlers
     #
     # :nocov:
     def setup!
-      Dir["#{Gruf.servers_path}/**/*.rb"].each do |f|
+      return unless File.directory?(controllers_path)
+      path = File.realpath(controllers_path)
+      $LOAD_PATH.unshift(path)
+      Dir["#{path}/**/*.rb"].each do |f|
+        next if f.include?('_pb') # exclude if people include proto generated files in app/rpc
```

Comment:
That is a pretty opinionated assumption based on naming convention - is it necessary? I'd prefer instead if we would inspect the file contents and only require it is it subclasses `::Gruf::Controllers::Base`. That would also protect us if someone includes controller concerns in the controller's directory - (see https://github.com/bigcommerce/gruf/pull/25)

Reply:
Yes - due to the way gRPC generates and includes files, we need to ensure this file is not included (otherwise it will error out with a "Service already defined" error). Unfortunately this is an issue due to the way gRPC generates the Ruby files, and not something in Gruf.

Reply:
Can we verify based on the contents of the file instead? That will allow us to ensure we only require Gruf::Controllers. The protobuf files won't be subclassing `::Gruf::Controllers::Base` so this method is the safest IMHO. The current loader will still load non-service files (such as ActiveSupport::Concerns) as mentioned in the previous PR.

Reply:
1) This doesn't _run_ or mount anything in those, which the prior did; it just loads the files.
2) We _can't_ load the file due to the require calls in the generated files, and I don't particularly want to have to do file opening and string parsing on the file itself.

Reply:
Ok. In our existing implementation I don't think we are using concerns anymore, but if this causes an issue I'll raise a new issue at that time.